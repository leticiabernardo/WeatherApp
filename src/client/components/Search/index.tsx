import { memo, useState } from 'react';
import {
  InputGroup,
  Input,
  InputLeftElement,
  InputRightElement,
  Button,
} from '@chakra-ui/react';
import { useAppContext } from '@/client/Context';
import { ReactComponent as IconSearch } from '@/assets/images/icon-search.svg';
import { ReactComponent as IconGeolocation } from '@/assets/images/icon-map.svg';

interface Props {
  fetchWeathers: (search: string) => void;
  setSearch: (search: string) => void;
}

const Search = ({ fetchWeathers, setSearch }: Props): JSX.Element => {
  const [searchValue, setSearchValue] = useState('');

  const handleSearchClick = () => {
    fetchWeathers(searchValue);
    setSearch(searchValue);
  };

  return (
    <div className="search">
      <InputGroup size="md">
        <InputLeftElement background="transparent">
          <Button
            h="1.75rem"
            size="sm"
            variant="link"
            color="white"
            _focus={{ outline: 'none' }}
          >
            <IconGeolocation fill="white" width="20px" height="20px" />
          </Button>
        </InputLeftElement>
        <Input
          placeholder="Nome da cidade..."
          outline="none"
          color="white"
          fontWeight="bold"
          _focus={{
            outline: 'none',
            backdropFilter: 'blur(9px)',
          }}
          _placeholder={{ color: '#a9a9a9' }}
          value={searchValue}
          onChange={e => setSearchValue(e.target.value)}
          onKeyPress={e => {
            if (e.key === 'Enter') {
              handleSearchClick();
            }
          }}
        />
        <InputRightElement width="3rem">
          <Button
            h="1.75rem"
            size="sm"
            variant="link"
            color="white"
            _focus={{ outline: 'none' }}
            onClick={handleSearchClick}
          >
            <IconSearch fill="white" width="20px" height="20px" />
          </Button>
        </InputRightElement>
      </InputGroup>
    </div>
  );
};

const MemoizedSearch = memo(Search, () => {
  return true;
});

const SearchWrapper = (): JSX.Element => {
  const { fetchWeathers, setSearch } = useAppContext();

  return <MemoizedSearch fetchWeathers={fetchWeathers} setSearch={setSearch} />;
};

export default SearchWrapper;
