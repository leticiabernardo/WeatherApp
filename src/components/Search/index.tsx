import { memo, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import {
  InputGroup,
  Input,
  InputLeftElement,
  InputRightElement,
  Button,
} from '@chakra-ui/react';
import { useAppContext } from '@/context/Context';
import { ReactComponent as IconSearch } from '@/assets/images/icon-search.svg';
import { ReactComponent as IconGeolocation } from '@/assets/images/icon-map.svg';

interface Props {
  fetchWeathers: (search: string) => void;
  setSearch: (search: string) => void;
  setError: (e: string) => void;
}

const Search = ({ fetchWeathers, setSearch, setError }: Props): JSX.Element => {
  const { t } = useTranslation();
  const [searchValue, setSearchValue] = useState('');

  const handleSearchClick = () => {
    if (searchValue === '') return;
    setSearch(searchValue);
  };

  const handleGeolocationClick = () => {
    if (!('geolocation' in navigator)) {
      setError(t('Geolocation is not supported by your browser'));
    }
    navigator.geolocation.getCurrentPosition(function (position): void {
      if (position?.coords?.latitude && position?.coords?.longitude) {
        const coords = [
          position.coords.latitude,
          position.coords.longitude,
        ].join(',');
        fetchWeathers(coords);
        setSearch(coords);
      }
    });
  };

  return (
    <div className="search">
      <InputGroup size="md">
        <InputLeftElement background="transparent">
          <Button
            h="1.75rem"
            size="sm"
            variant="link"
            color="white"
            _focus={{ outline: 'none' }}
            onClick={handleGeolocationClick}
          >
            <IconGeolocation fill="white" width="20px" height="20px" />
          </Button>
        </InputLeftElement>
        <Input
          placeholder={t('city, state or country...')}
          outline="none"
          color="white"
          fontWeight="bold"
          _focus={{
            outline: 'none',
            backdropFilter: 'blur(9px)',
          }}
          _placeholder={{ color: '#a9a9a9' }}
          value={searchValue}
          onChange={e => setSearchValue(e.target.value)}
          onKeyPress={e => {
            if (e.key === 'Enter') {
              handleSearchClick();
            }
          }}
        />
        <InputRightElement width="3rem">
          <Button
            h="1.75rem"
            size="sm"
            variant="link"
            color="white"
            _focus={{ outline: 'none' }}
            onClick={handleSearchClick}
          >
            <IconSearch fill="white" width="20px" height="20px" />
          </Button>
        </InputRightElement>
      </InputGroup>
    </div>
  );
};

const MemoizedSearch = memo(Search, () => {
  return false;
});

const SearchWrapper = (): JSX.Element => {
  const { fetchWeathers, setSearch, setError, search } = useAppContext();

  useEffect(() => {
    fetchWeathers(search);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [search]);

  return (
    <MemoizedSearch
      fetchWeathers={fetchWeathers}
      setSearch={setSearch}
      setError={setError}
    />
  );
};

export default SearchWrapper;
