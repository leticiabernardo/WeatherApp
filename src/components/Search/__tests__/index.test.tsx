import { render, cleanup, fireEvent } from '@testing-library/react';
import Search from '../index';

const props = {
  setSearch: jest.fn(),
  setError: jest.fn(),
};

describe('search render', () => {
  afterEach(cleanup);

  test('render the component', () => {
    const { container } = render(<Search {...props} />);
    expect(container).toMatchSnapshot();
  });
});

describe('search component events', () => {
  afterEach(cleanup);

  test('should click button to search with empty value', async () => {
    const { container, getByTestId } = render(<Search {...props} />);
    const button = getByTestId('search-button');
    await fireEvent.click(button);
    expect(container).toMatchSnapshot();
  });

  test('should change the value of the search field', async () => {
    const { findByPlaceholderText } = render(<Search {...props} />);
    const input = (await findByPlaceholderText(
      'city, state or country...'
    )) as HTMLInputElement;
    await fireEvent.change(input, { target: { value: 'ceara' } });

    expect(input.value).toBe('ceara');
  });

  test('should click button to search value', async () => {
    const { container, getByTestId, findByPlaceholderText } = render(
      <Search {...props} />
    );
    const input = await findByPlaceholderText('city, state or country...');
    await fireEvent.change(input, { target: { value: 'ceara' } });

    const button = getByTestId('search-button');
    await fireEvent.click(button);
    expect(container).toMatchSnapshot();
  });

  test('should key enter press to search value', async () => {
    const { container, findByPlaceholderText } = render(<Search {...props} />);

    const input = await findByPlaceholderText('city, state or country...');
    await fireEvent.change(input, { target: { value: 'ceara' } });
    fireEvent.keyPress(input, { key: 'Enter', code: 13, charCode: 13 });
    expect(container).toMatchSnapshot();
  });
});

describe('successfully render location (latitude and longitude) by browser', () => {
  afterEach(cleanup);

  beforeEach(() => {
    const mockGeolocation = {
      getCurrentPosition: jest.fn().mockImplementationOnce(success =>
        Promise.resolve(
          success({
            coords: {
              latitude: 51.1,
              longitude: 45.3,
            },
          })
        )
      ),
      clearWatch: jest.fn(),
      watchPosition: jest.fn(),
    };

    const navigator = { geolocation: mockGeolocation };

    Object.defineProperty(window, 'navigator', {
      value: navigator,
      writable: true,
    });
  });

  test('should click the button to get the users geolocation', async () => {
    const { container, getByTestId } = render(<Search {...props} />);
    const button = getByTestId('geolocation-button');
    await fireEvent.click(button);
    expect(container).toMatchSnapshot();
  });
});

describe('error to render the location (latitude and longitude) by browser', () => {
  afterEach(cleanup);

  beforeEach(() => {
    const navigator = {};

    Object.defineProperty(window, 'navigator', {
      value: navigator,
      writable: true,
    });
  });

  test('should return error when clicking button to get user location', async () => {
    const { getByTestId } = render(<Search {...props} />);
    const button = getByTestId('geolocation-button');
    await fireEvent.click(button);
  });
});
