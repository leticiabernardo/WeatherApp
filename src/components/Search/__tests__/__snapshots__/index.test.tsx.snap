// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search component events should click button to search value 1`] = `
<div>
  <div>
    <div
      class="chakra-input__group css-4302v8"
    >
      <div
        class="chakra-input__left-element css-5tdink"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="geolocation-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
      <input
        class="chakra-input css-ryoawo"
        placeholder="city, state or country..."
        value="ceara"
      />
      <div
        class="chakra-input__right-element css-jwi0po"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="search-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`search component events should click button to search with empty value 1`] = `
<div>
  <div>
    <div
      class="chakra-input__group css-4302v8"
    >
      <div
        class="chakra-input__left-element css-5tdink"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="geolocation-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
      <input
        class="chakra-input css-ryoawo"
        placeholder="city, state or country..."
        value=""
      />
      <div
        class="chakra-input__right-element css-jwi0po"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="search-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`search component events should key enter press to search value 1`] = `
<div>
  <div>
    <div
      class="chakra-input__group css-4302v8"
    >
      <div
        class="chakra-input__left-element css-5tdink"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="geolocation-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
      <input
        class="chakra-input css-ryoawo"
        placeholder="city, state or country..."
        value="ceara"
      />
      <div
        class="chakra-input__right-element css-jwi0po"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="search-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`search render render the component 1`] = `
<div>
  <div>
    <div
      class="chakra-input__group css-4302v8"
    >
      <div
        class="chakra-input__left-element css-5tdink"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="geolocation-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
      <input
        class="chakra-input css-ryoawo"
        placeholder="city, state or country..."
        value=""
      />
      <div
        class="chakra-input__right-element css-jwi0po"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="search-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`successfully render location (latitude and longitude) by browser should click the button to get the users geolocation 1`] = `
<div>
  <div>
    <div
      class="chakra-input__group css-4302v8"
    >
      <div
        class="chakra-input__left-element css-5tdink"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="geolocation-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
      <input
        class="chakra-input css-ryoawo"
        placeholder="city, state or country..."
        value=""
      />
      <div
        class="chakra-input__right-element css-jwi0po"
      >
        <button
          class="chakra-button css-dzjm3n"
          data-testid="search-button"
          type="button"
        >
          <div
            class="svg-mock"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
